### PR TITLE
Adds caching decorator.

### PR DESCRIPTION
**Pull request by bsiemon**

### CODE REVIEW

File: tests/test_cache.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+"""
+Tests for cache.py
+"""
+from upside_core.cache import InMemoryCache, cached_get, cache_update
+
+
+class CacheTestService(object):
+
+    def __init__(self):
+        self.cache = InMemoryCache()
```
Comment (janrubio) [inline tests/test_cache.py]: Should we add a test case for passing this as the `cache_factory` to the decorator?

Comment (bsiemon) [inline tests/test_cache.py]: good catch, I will add that today if I can.

File: upside_core/cache.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    into the cache.
+
+    The item at arg[0] should be the 'self' param for the class instance that has
+    an instance of an Cache client at attr: 'cache'
+
+    OR
+
+    the function may have a keyword arg: 'cache_factory' that returns a cache instance if called.
+
+    :param create_key: A function ref that when called with the argument at index 1 returns
```
Comment (janrubio) [inline upside_core/cache.py]: What do you think of passing this function all the arguments? In the case of the merchant service processor retrieval, the key could be the `site_uuid` which is index 2:

```python
def get_processors_by_site(self, merchant_id, site_id, **kwargs):
```
The cache key for the list could look like this: `merchant-service:site:{site_uuid}:processors`

Comment (bsiemon) [inline upside_core/cache.py]: good point, I will do that.

File: upside_core/cache.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+    def my_extractor(result):
+        return (result['key'], result['part'], result['id'])
+
+    @cache_update(extract_cache_item=my_extrator
+    def my_mutator(self, arg1, arg2):
+        # Do work
+
+        return {
+            'key': 'my-cache-key',
```
Comment (janrubio) [inline upside_core/cache.py]: Do we care that knowledge of how this key is constructed is within the `my_mutator` itself as opposed to how it's done in `cached_get`?

**Pull request merged**